### PR TITLE
Simpler iterator interface

### PR DESCRIPTION
**Pull request by watson28**
solves https://github.com/spirit-labs/tektite/issues/82

### CODE REVIEW

**Comment by purplefox** [comment]
Hi @watson28 thanks for this!

There are a few conflicts after I merged the work on refactoring store. The store stuff has mainly moved to proc/storage.go

**Comment by watson28** [comment]
@purplefox I will wait for the merge of https://github.com/spirit-labs/tektite/pull/142 since it has additional changes to iterators.

**Comment by purplefox** [comment]
It's merged now. Hopefully won't be changes to iterators for a while :)

**Comment by watson28** [comment]
@purplefox, conflicts are resolved.

**Comment by purplefox** [comment]
Since MergingIterator is performance sensitive code, it would nice to see a benchmark which iterates through a bunch of keys using MergingIterator and compares performance before and after the changes.

File: iteration/iter.go [hunk 1]
```diff
@@ -1,10 +1,10 @@
...
-import "github.com/spirit-labs/tektite/common"
+import (
+	"github.com/spirit-labs/tektite/common"
+)
 
 type Iterator interface {
-	Current() common.KV
-	Next() error
-	IsValid() (bool, error)
+	Next() (bool, common.KV)
```
[the remark this answers is not in the document]
Comment (watson28) [reply iteration/iter.go:8]: makes senses @purplefox . I will change the signature to be:

```suggestion
	Next() (bool, common.KV) (bool, common.KV, error)
```

**Comment by watson28** [comment]
@purplefox I addressed your comment:
-  I changed the Next() method to return an error
- I moved the logic in `dropNextKeysEqualsToLatest` to `readIterHeadOrNext`

Regarding:
>Since MergingIterator is performance-sensitive code, it would be nice to see a bench iterates through a bunch of keys using MergingIterator and compares performance before and after the changes.

It seems we have `merging_iter_bench_test.go` for this. I ran the benchmark, and thats is what I got:

This branch:
```
goos: darwin
goarch: arm64
pkg: github.com/spirit-labs/tektite/iteration
BenchmarkMergingIterator-10    	    5548	    183379 ns/op	   64355 B/op	    1003 allocs/op
PASS
ok  	github.com/spirit-labs/tektite/iteration	1.262s
```

main branch:
```
goos: darwin
goarch: arm64
pkg: github.com/spirit-labs/tektite/iteration
BenchmarkMergingIterator-10    	    3578	    305500 ns/op	     296 B/op	       3 allocs/op
PASS
ok  	github.com/spirit-labs/tektite/iteration	1.278s
```

It seems we now have an improvement in the "Nanoseconds per operation" (third column) but now we are allocating more memory (last two columns).

I will work on refactoring the code to consume less memory.

**Comment by purplefox** [comment]
I wonder if maintaining `iterHeads` is really necessary?

**Comment by purplefox** [comment]
> I will work on refactoring the code to consume less memory.

Ack, it should be possible to avoid creating new structs in the loops (kvv etc)

**Comment by watson28** [comment]
@purplefox 

this is the new benchmark with the latest changes

```
goos: darwin
goarch: arm64
pkg: github.com/spirit-labs/tektite/iteration
BenchmarkMergingIterator-10    	    6270	    189081 ns/op	     277 B/op	       3 allocs/op
PASS
```
Now we have better numbers in "Nanoseconds per operation" and "Number of bytes allocated per operation".

basically, the increase in memory was caused by the new functions  I created to improve the readability of the code (`readIterHead` and the 3 functions to log messages). New function calls means more memory stack allocated.  I inlined the code in the function as it originally was, and the memory usage was reduced.

The other change needed was to remove the `iterHeads` array. However, to do this, I have to bring back the `Current()` method to the iterator interface. This is needed in `merging_iter` since we have to keep a reference to the latest iter.Next() that is returned, in case that is not used in the current Next() call of `merging_iter`. Taking this example for the unit test:

```
it1: [2, 5, 9]
it2: [0, 7, 10]
it3: [4, 6, 8]
```

calling `mergin_iter.Next()` will call next() in the 3 iterators and choose the smallest key, it2 -> 0. But then we have advanced it1 and it3. so we need to have a way to keep the latest value of each item. Otherwise, we will lose the values `2` and `4` in the `Next()` call of `merging_iter`. That was achieved with `iterHeads` and now with the `Current()` method in the iterators.

**Comment by purplefox** [comment]
Thanks! I'll give this a proper look tomorrow morning.

**Comment by purplefox** [comment]
Just a thought: If we're going to re-introduce `Current()`, then it seems redundant for `Next()` to return a KV. I would suggest it should just return `(bool, error)`

**Comment by purplefox** [comment]
Other than previous comment, LGTM

**Comment by watson28** [comment]
>Just a thought: If we're going to re-introduce Current(), then it seems redundant for Next() to return a KV. I would suggest it should just return (bool, error)

@purplefox After looking at the code base, I noticed that most of the uses of iterators, except for `merging_iter,` satisfy their needs by only calling `Next()` and getting the value from the return value. It feels weird to change these and replace a single call with two function calls only because `merging_iter` needs it. I think the proper iterator interface should have the `Next()` method return the value and still provide `Current()` for those few cases where caching the last Next() Value is not an option for the client(`mergin_iter` case).

Let me know what you think.

**Comment by purplefox** [comment]
> Let me know what you think.

That's reasonable. I will merge.

Thanks for your efforts! Great piece of work

**Comment by purplefox** [comment]
I've had to revert this PR for now, as it seems to cause the continuous CI run to fail, e.g. https://github.com/spirit-labs/tektite/actions/runs/9896394329

**Comment by purplefox** [comment]
I would suggest running the failover integration tests in a loop locally. (There is a script `run_tests_in_loop.sh` in the root of the project which can help with that.
From past experience I would guess the problem is something related to the minCompactableVersion logic in the merging iterator, as this logic is exercised during failover.

**Comment by purplefox** [comment]
One way to see if there any behavioural differences between the old merging iterator and the new one (and this will probably be easier than debugging the failover tests on CI), would be to use a form of fuzz testing.

In a loop, generate a bunch of random keys and similar keys, with same and different versions and then run them through both the old and new merging iterators and verify that the output is the same.

**Comment by purplefox** [comment]
@watson28 Just to keep you informed - I have found a couple of bugs and will submit a patch in the next day or so.

**Comment by watson28** [comment]
>@watson28 Just to keep you informed - I have found a couple of bugs and will submit a patch in the next day or so.

@purplefox, thanks for reapplying the changes and the patch with the fix.

**Pull request merged**